cpp/src/copying/copy.cpp

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/cudf.h>
+#include <cudf/copying.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <algorithm>
+
+namespace cudf
+{
+namespace experimental
+{
+namespace detail
+{
+
+inline mask_state should_allocate_mask(mask_allocation_policy mask_alloc, bool mask_exists) {
+  if ((mask_alloc == ALWAYS) || (mask_alloc == RETAIN && mask_exists)) {
+    return UNINITIALIZED;
+  } else {
+    return UNALLOCATED;
+  }
+}
+
+/*
+ * Initializes and returns an empty column of the same type as the `input`.
+ */
+std::unique_ptr<column> empty_like(column_view input, cudaStream_t stream)
+{
+  std::vector<std::unique_ptr<column>> children {};
+  children.reserve(input.num_children());
+  for (size_type index; index < input.num_children(); index++) {
+      children.emplace_back(empty_like(input.child(index), stream));
+  }
+
+  return std::make_unique<column>(input.type(), 0, rmm::device_buffer {},
+		                  rmm::device_buffer {}, 0, std::move(children));
+}
+
+/*
+ * Creates an uninitialized new column of the same size and type as the `input`.
+ * Supports only fixed-width types.
+ */
+std::unique_ptr<column> allocate_like(column_view input,
```

Comment:
To reduce duplicated code, this API should just call the other `allocate_like` and pass in `input.size()` for the size parameter.